0008-zkvk-keccak.md

```diff
@@ -0,0 +1,665 @@
+    constrain( word(x) * 2^1 - ( quotient(x) * 2^64 + remainder(x)) )
+    constrain( rotated(x) - (quotient(x) + remainder(x)) )
+    constrain( bound(x) - (quotient(x) + 2^64 - 2^1) )
+
+for x in [0..5)
+    for i in [0..4)
+        constrain( state_d(x)[i] - (reset0_c(x-1)[i] + expand_rot_c(x+1)[1]) )
+```
+
+and $180$ lookups.
+
+
+Next, for each row `x in [0..5)` and column `y in [0..5)`, compute (with no need of resetting `D`):
+
+$$
+\begin{align*}
+E[x][y] := &\ A[x][y]\ \oplus\ D[x] \\
+\iff \\
+sparse(E[x][y]) := &\ sparse(A[x][y]) + sparse(D[x])\\
+\end{align*} 
+$$
+
+This part uses the following 100 constraints
+
+```rust
+for i in [0..4)
+    for x in [0..5)
+        for y in [0..5)
+            constrain( state_e(x,y)[i] - (state_a(x,y)[i] + state_d(x)[i]) )
+```
+
+| Columns: | [340...440) |
+| -------- | ----------- |
+| Theta    | state_e     |
+
+#### Step pi-rho
+
+For each row `x in [0..5)` and column `y in [0..5)`, perform (into quarters):
+
+$$
+\begin{align*}
+B[y][2x+3y] := &\ ROT(E[x][y], OFF[x][y]) \\
+\iff \\
+expand(B[y][2x+3y]) := &\ ROT(expand(E[x][y]), OFF[x][y])
+\end{align*} 
+$$
+
+| Columns: | [440...840) | [840...940) | [940...1040) | [1040...1140) | [1140...1240) | [1240...1340) | [1340...1440) | [1440...1540) | 
+| -------- | ----------- | ----------- | ------------ | ------------- | ------------- | ------------- | ----- | ---- |
+| PiRho    | reset_e     | dense_e     | quotient_e   | remainder_e   | bound_e       | dense_rot_e   | expand_rot_e | state_b |
+
+Recall that in order to perform the rotation operation, the state needs to be reset. This step can be carried out with the following $125$ constraints and $800 (=400+100+100+100+100)$ lookups:
+
+```rust
+for x in [0...5)
+    for y in [0...5)
+        constrain( state_e[i] - (reset0_e(x,y)[i] + 2*reset1_e(x,y)[i] + 4*reset2_e(x,y)[i] + 8*reset3_e(x,y)[i] ) )
+        let word(x,y)      = dense_e(x,y)[0]     + 2^16*dense_e(x,y)[1]     + 2^32*dense_e(x,y)[2]     + 2^48*dense_e(x,y)[3]
+        let quotient(x,y)  = quotient_e(x,y)[0]  + 2^16*quotient_e(x,y)[1]  + 2^32*quotient_e(x,y)[2]  + 2^48*quotient_e(x,y)[3]
+        let remainder(x,y) = remainder_e(x,y)[0] + 2^16*remainder_e(x,y)[1] + 2^32*remainder_e(x,y)[2] + 2^48*remainder_e(x,y)[3]
+        let bound(x,y)     = bound_e(x,y)[0]     + 2^16*bound_e(x,y)[1]     + 2^32*bound_e(x,y)[2]     + 2^48*bound_e(x,y)[3]
+        let rotated(x,y)   = dense_rot_e(x,y)[0] + 2^16*dense_rot_e(x)[1] + 2^32*dense_rot_e(x,y)[2] + 2^48*dense_rot_e(x,y)[3]
+        constrain( word(x,y) * 2^(OFF(x,y)) - ( quotient(x) * 2^64 + remainder(x)) )
+        constrain( rotated(x,y) - (quotient(x,y) + remainder(x)) )
+        constrain( bound(x,y) - (quotient(x,y) + 2^64 - 2^(OFF(x,y)) )
+        constrain( state_b(y,2x+3y) - expand_rot_e(x,y) )
+```
+
+Note that the value `OFF(x,y)` will be a different constant for each index, according to the rotation table of Keccak.
```

Comment:
How/where is this encoded?

Reply:
I thought it would be encoded directly onto the constraints themselves